feat(integrations): add salesforce integration

pashafateev opened the pull request:
**testing**: successfully authenticated via OAuth and made an API call using our Python SDK for both an external client app and a connected app.

Refs: INT-54

pashafateev committed:
init

itayd committed:
feat(sessionsrpc): straight input of a json object (#1054)

instead of ingesting param by param which presents difficulties with
jsonifying parameters, just accept a straight raw json object which must
be a dictionary and process it as an input. all the "smart" conversion
is done on the backend side.

---------

Co-authored-by: Daniel Abraham <daniel@autokitteh.com>

daabr committed:
fix(zoom): oauth flow (#1056)

The original scopes I specified were legacy-style, which is no longer
allowed for new apps. Need to use only new granular scopes.

Anyway, passing an empty list is fine because it uses the apps default
scopes, and it's actually preferable because there are 100+ granular
scopes that we currently use.

Tested that an OAuth 3-legged flow really works, documented in
https://docs.autokitteh.com/integrations/zoom

Refs: INT-241

tebeka committed:
fix(python): Better error from activity reply (#1049)

Fixes: ENG-1990

daabr committed:
refactor(integrations): webhook handlers (#1055)

New reusable functions to register UI/save/OAuth webhook handlers in
`integrations/common/server.go` - mostly for new-style integrations, but
not just.

Most of the PR is tweaks related to that (using these new functions), or
cosmetic tweaks (e.g. rename `muxes *muxes.Muxes` variable to `m` to be
more idiomatic about Go var naming).

mario99logic committed:
fix: Block connection deletion with active tiggers (#1035)

Implemented `doesConnectionHaveTriggers` function to check for existing triggers and
integrated it into `deleteConnectionsAndVars`.

daabr committed:
feat(zoom): support server-to-server auth (#1059)

Tested only the saving of credentials, not their usage - that should be
done in INT-247.

Refs: INT-241

mario99logic committed:
fix: redirect from Github marketplace (#1057)

Marketplace Installation Flow Enhancement

- Added handleGitHubMarketplace function to handle direct app
installations from GitHub Marketplace
- Redirects users to GitHub app installations page after successful
installation
- Tested functionality using ngrok for local development

The redirect currently points to
`/settings/installations/{installation_id}`. We could alternatively
redirect to `/apps/installations/{installation_id}` to maintain
marketplace context.

Refs: INT-172

daabr committed:
fix(integrations): don't prepend "connect" to legacy webhooks (#1060)

Regression caused by #1055.

Solution: don't use `LegacyDescriptor` (which added `/connect/` after
the integration's URL path prefix, and move all relevant static web
content accordingly (`xxx/connect/*` --> `xxx/*`).

Tested: Google Sheets, AWS, Asana, Slack.

Also removed the obsolete "HTTP" connection UI - replaced by webhook
triggers.

Refs: INT-257

itayd committed:
feat(py-sdk): oauth2 requests session for height and zoom (#1058)

efiShtain committed:
use hard delete and detach foreign keys for sessions and events only (#992)

this is a simplified version of this PR
https://github.com/autokitteh/autokitteh/pull/965

it handles only sessions (and its dependents) and events

deleting trigger / connection does not delete events but remove the
foreign key from events on trigger/connection

deleting event does not delete sessions but remove the foreign key from
session on events

deleting session delete cascade the call attempts, log records and call
specs

itayd committed:
feat(sessions): get history from temporal, not from db (#1062)

motivation:
- do not do extra activities on any single activity, saves history
length and size.
- be more aligned with temporal to ease future handling of concurrency
and retries.

changes:
- instead of getting workflow history from the db, get it directly from
temporal, translating its history into autokitteh history.
- prints go to the db, and are reported only on GetPrints
- print now can report any arbitrary value, but currently implemented
only string

notes:
- this will only show actual temporal activities in the history for
calls, so next_event, sleep etc would not show.
- for migration purposes (deploying while workflows are running), old
workflows would still be persisting data to the log via activities, but
all new ones will not.

daabr committed:
fix(github): static UI shouldn't override webhooks (#1064)

Refs: INT-257, #1055, #1060

itayd committed:
feat(dispatcher): use workflow.SendExternalSignal instead of using the temporal client (#1061)

i guess it links the workflows as well in the temporal level somehow. no
need to wrap signalworkflow in an activity - so less activites

daabr committed:
fix(google): forgot invalid gmail watches (#1065)

Similar to how we stop watching invalid Google Forms watches.

This PR refactors the Google Forms code to be more generic for all
Google watches, even though for now we add this only for Gmail.

This fixes a recurring issue in production, related to inactive users:


https://app.datadoghq.com/logs?query=env%3Astaging%20status%3Aerror&agg_m=count&agg_m_source=base&agg_t=count&clustering_pattern_field_path=message&cols=source%2Chost%2Cservice&event=AwAAAZTz5YalL1_nigAAABhBWlR6NVlwTUFBQW5RWTFEU0x2WkRRQUIAAAAkMDE5NGYzZTUtOTY3MS00MTY1LWE3YTctMzgyNGYwZmEwNjU1AAAAag&fromUser=true&messageDisplay=inline&refresh_mode=sliding&storage=hot&stream_sort=desc&viz=stream&from_ts=1739255096534&to_ts=1739258696534&live=true

daabr committed:
chore: update deps & web UI (#1066)

daabr committed:
refactor(integrations): merge client.go and server.go (#1067)

Only new integrations: Height, Linear, MS, MS Teams, Zoom.

Purely cosmetic, no functional change, but it's a small step in the path
to refactor integrations and their API, to have simpler templates.

itayd committed:
chore: remove no longer needed migration code for users (#1069)

efiShtain committed:
add indexes on events / sessions (#1071)

we need to query events and sessions by created_at field so we can
delete them in a fast way without scanning the entire table

mario99logic committed:
feat: Add noctx linter (#1070)

Added noctx linter and fixed context propagation issues

Changes:
- Added noctx linter to golangci.yaml to detect missing context
propagation in HTTP requests
- Fixed context propagation in integrations (Asana, Jira, Google, Auth0,
HubSpot) by:
  - Using http.NewRequestWithContext instead of http.NewRequest
  - Propagating context from HTTP handlers using r.Context()
  - Replacing http.Client.Get with proper context-aware requests
- Added context propagation in backend services where needed using
context.Background()

Testing:
- Manually tested by creating connections for affected integrations
- Verified connection tests (connTest) work correctly with new context
changes
- Would appreciate guidance on testing strategy for backend context
changes, specifically for internal/backend/usagereporter/dataposter.go
if needed

ref: INT-191

pashafateev committed:
add salesforce integration

pashafateev committed:
add salesforce integration

pashafateev committed:
add logo

daabr commented on web/static/salesforce/styles.css:
```diff
@@ -0,0 +1,80 @@
+  .submit-btn:hover {
+    background-color: #2980b9;
+  }
+  
```
Remove whitespaces, so the file ends in `\n`
```suggestion

```

daabr commented on web/static/salesforce/styles.css:
```diff
@@ -0,0 +1,80 @@
+body {
```
Please autoformat this file with Prettier

pashafateev replied:
done

pashafateev replied:
Done

daabr commented on web/static/salesforce/index.html:
```diff
@@ -0,0 +1,108 @@
+      <div class="form-group">
+        <select name="auth_type" id="authType">
+          <option value="oauthPrivate">Private OAuth 2.0</option>
+        </select>
```
If there's no user choice, we might as well make it a pre-populated read-only field (and hidden in the official web UI, but that's beyond the scope of this PR).

```suggestion
        <input type="text" name="auth_type" value="oauthPrivate" readonly />
```

daabr commented on web/static/salesforce/index.html:
```diff
@@ -0,0 +1,108 @@
+      <div id="oauthSection">
+        <h2>OAuth 2.0</h2>
+      </div>
+
```
No need to specify in this internal UI an unused section.
```suggestion
```

daabr commented on web/static/salesforce/index.html:
```diff
@@ -0,0 +1,108 @@
+        <h2>OAuth 2.0</h2>
+      </div>
+
+      <div id="privateOauthSection">
```
No need for an ID if we're not manipulating this section with JavaScript.
```suggestion
      <div>
```

daabr commented on web/static/salesforce/index.html:
```diff
@@ -0,0 +1,108 @@
+          <label for="client_id">Client ID <i>(required)</i></label>
+          <input
+            type="text"
+            id="clientId"
```
```suggestion
```

daabr commented on web/static/salesforce/index.html:
```diff
@@ -0,0 +1,108 @@
+          <label for="client_secret">Client Secret <i>(required)</i></label>
+          <input
+            type="text"
+            id="clientSecret"
```
```suggestion
```

daabr commented on web/static/salesforce/index.html:
```diff
@@ -0,0 +1,108 @@
+          <label for="instance_url">Instance URL <i>(required)</i></label>
+          <input
+            type="text"
+            id="instanceUrl"
```
```suggestion
```

daabr commented on web/static/images/salesforce.png:
Instead of this PNG, use this:
- https://en.m.wikipedia.org/wiki/File:Salesforce.com_logo.svg
- https://upload.wikimedia.org/wikipedia/commons/f/f9/Salesforce.com_logo.svg

And also mention these links in `images/sources.txt`.

pashafateev replied:
Done

daabr commented on web/static/salesforce/index.html:
```diff
@@ -0,0 +1,108 @@
+
+  <body>
+    <a href="/" target="_self">
+      <img class="banner" src="/static/banner.svg" alt="AutoKitteh" />
```
Broken link - you're using a PNG image, not SVG. But anyway see my comment about that image file.

pashafateev replied:
Done

daabr commented on web/static/salesforce/index.html:
```diff
@@ -0,0 +1,108 @@
+
+        <div class="form-group">
+          <label for="instance_url">Instance URL <i>(required)</i></label>
+          <input
```
Nit: if you auto-format this file, I think Prettier will remove the line breaks from this element because it's short enough to fit in a single line.

pashafateev replied:
It did not.

daabr commented on web/static/salesforce/form.js:
You're not using JS, so remove this file (and also remove line 25 in `index.html`).

daabr commented on runtimes/pythonrt/py-sdk/pyproject.toml:
```diff
@@ -28,6 +28,7 @@ all = [
 	"msgraph-sdk ~= 1.18",
 	"openai ~= 1.57",
 	"PyGithub @ git+https://github.com/PyGithub/PyGithub.git@b5f8f078f2b7fc0db6a041499c7a05b0bc7e750d",
+	"simple-salesforce ~= 1.12",
```
Also add this in `/pythonrt/runner/pyproject.toml`.

pashafateev replied:
Done

daabr commented on internal/backend/oauth/oauth.go:
```diff
@@ -475,6 +475,18 @@ func New(l *zap.Logger, vars sdkservices.Vars) sdkservices.OAuth {
 					// "Teamwork.Migrate.All", // Application-only.
 				},
 			},
+			"salesforce": {
```
Do you have a better link?
```suggestion

            // https://help.salesforce.com/s/articleView?id=xcloud.remoteaccess_oauth_flows.htm
			// https://help.salesforce.com/s/articleView?id=xcloud.remoteaccess_oauth_endpoints.htm
			"salesforce": {
```

pashafateev replied:
Done

daabr commented on internal/backend/oauth/oauth.go:
```diff
@@ -475,6 +475,18 @@ func New(l *zap.Logger, vars sdkservices.Vars) sdkservices.OAuth {
+					TokenURL: "https://login.salesforce.com/services/oauth2/token",
+				},
+				RedirectURL: redirectURL + "salesforce",
+				Scopes:      []string{"full"},
```
```suggestion
				// https://help.salesforce.com/s/articleView?id=xcloud.remoteaccess_oauth_tokens_scopes.htm
				Scopes: []string{"full"},
```

daabr commented on integrations/salesforce/client.go:
```diff
@@ -0,0 +1,39 @@
+			SupportsConnectionTest: true,
+		},
+	}))
+)
```
```suggestion
var desc = common.Descriptor("salesforce", "Salesforce", "/static/images/salesforce.svg")
```

daabr commented on integrations/salesforce/client.go:
```diff
@@ -0,0 +1,39 @@
+package salesforce
+
+import (
+	"go.autokitteh.dev/autokitteh/integrations/microsoft/connection"
```
Microsoft?

pashafateev replied:
Done

daabr commented on integrations/salesforce/client.go:
```diff
@@ -0,0 +1,39 @@
+func New(v sdkservices.Vars, o sdkservices.OAuth) sdkservices.Integration {
+	return sdkintegrations.NewIntegration(
+		desc, sdkmodule.New(),
+		connection.Status(v), connection.Test(v, o),
```
```suggestion
		desc, sdkmodule.New(), status(v), test(v, o),
```

daabr commented on integrations/salesforce/server.go:
```diff
@@ -0,0 +1,42 @@
+package salesforce
```
Please append this file to `client.go` and rename to `integration.go` (see #1067)

pashafateev replied:
Done

daabr commented on integrations/salesforce/checks.go:
```diff
@@ -0,0 +1,43 @@
+	})
+}
+
+// TODO: Implement this.
```
```suggestion
// test checks whether the connection is actually usable, i.e. the configured
// authentication credentials are valid and can be used to make API calls.
// TODO: Implement this.
```

daabr commented on integrations/salesforce/checks.go:
```diff
@@ -0,0 +1,43 @@
+	})
+}
+
+// TODO: Implement this.
```
Add Linear ticket ID

pashafateev replied:
Done

daabr commented on integrations/salesforce/checks.go:
```diff
@@ -0,0 +1,43 @@
+	"go.autokitteh.dev/autokitteh/sdk/sdktypes"
+)
+
+// Status checks the connection's initialization status (is it
```
The first word in Go function docs needs to be the function name, and it needs to match even if it's lower-case.
```suggestion
// status checks the connection's initialization status (is it
```

daabr commented on integrations/salesforce/server.go:
```diff
@@ -0,0 +1,42 @@
+)
+
+// Start initializes all the HTTP handlers of all the Salesforce integrations. This
+// includes connection UIs, connection initialization webhooks, and event webhooks.
```
Recent change in #1067
```suggestion
// Start initializes all the HTTP handlers of the integration.
// This includes an internal connection UI, webhooks for AutoKitteh
// connection initialization, and asynchronous event webhooks.
```

daabr commented on integrations/salesforce/server.go:
```diff
@@ -0,0 +1,42 @@
+
+// Start initializes all the HTTP handlers of all the Salesforce integrations. This
+// includes connection UIs, connection initialization webhooks, and event webhooks.
+func Start(l *zap.Logger, muxes *muxes.Muxes, v sdkservices.Vars, o sdkservices.OAuth, d sdkservices.DispatchFunc) {
```
```suggestion
func Start(l *zap.Logger, m *muxes.Muxes, v sdkservices.Vars, o sdkservices.OAuth, d sdkservices.DispatchFunc) {
```

daabr commented on integrations/salesforce/server.go:
```diff
@@ -0,0 +1,42 @@
+func Start(l *zap.Logger, muxes *muxes.Muxes, v sdkservices.Vars, o sdkservices.OAuth, d sdkservices.DispatchFunc) {
+	// Connection UI for authenticated AutoKitteh users (user authentication
+	// isn't required, but it makes no sense to create a connection without it).
+	muxes.Auth.Handle("GET /salesforce/", http.FileServer(http.FS(static.SalesforceWebContent)))
```
```suggestion
	common.ServeStaticUI(m, desc, static.SalesforceWebContent)
```

daabr commented on integrations/salesforce/server.go:
```diff
@@ -0,0 +1,42 @@
+	h := newHTTPHandler(l, v, o, d)
+	muxes.Auth.HandleFunc("POST /salesforce/save", h.handleSave)
+	muxes.Auth.HandleFunc("GET /salesforce/save", h.handleSave)
+	muxes.Auth.HandleFunc("GET /salesforce/oauth", h.handleOAuth)
```
```suggestion
	h := newHTTPHandler(l, v, o, d)
	common.RegisterSaveHandler(m, desc, h.handleSave)
	common.RegisterOAuthHandler(m, desc, h.handleOAuth)
```

daabr committed:
fix(gmail): handle watch renewal failures gracefully (#1073)

1. If we determine that a Gmail connection's auth is revoked, and forget
its watch, don't retry to renew it every 8 hours
2. Don't log an error for every watch renewal failure, only ones that
are really errors that require a retry (which is done by Temporal
anyway)

Refs: INT-264

pashafateev committed:
Update web/static/salesforce/styles.css

Co-authored-by: Daniel Abraham <daniel@autokitteh.com>

pashafateev committed:
Update integrations/salesforce/checks.go

Co-authored-by: Daniel Abraham <daniel@autokitteh.com>

daabr commented on integrations/salesforce/vars.go:
```diff
@@ -0,0 +1,36 @@
+}
+
+// privateOAuth contains the user-provided details of
+// a private Linear OAuth 2.0 app.
```
```suggestion
// privateOAuth contains the user-provided details of a private Salesforce OAuth 2.0 app.
```

pashafateev committed:
Apply suggestions from code review

Co-authored-by: Daniel Abraham <daniel@autokitteh.com>

pashafateev committed:
Apply suggestions from code review

Co-authored-by: Daniel Abraham <daniel@autokitteh.com>

pashafateev committed:
Apply suggestions from code review

Co-authored-by: Daniel Abraham <daniel@autokitteh.com>

pashafateev committed:
address code review comments

pashafateev committed:
Merge branch 'main' into pasha/int-54

pashafateev committed:
run gen-reqs.py

daabr commented on integrations/salesforce/checks.go:
```diff
@@ -0,0 +1,45 @@
+// TODO(INT-268): Implement this.
+func test(v sdkservices.Vars, o sdkservices.OAuth) sdkintegrations.OptFn {
+	return sdkintegrations.WithConnectionTest(func(ctx context.Context, cid sdktypes.ConnectionID) (sdktypes.Status, error) {
+		return sdktypes.NewStatus(sdktypes.StatusCodeUnspecified, "Not implemented"), nil
```
`Unspecified` in proto enums is like "null", it should never be used explicitly, use `Error` instead.

pashafateev replied:
Done

daabr commented on internal/backend/oauth/oauth.go:
```diff
@@ -476,6 +476,22 @@ func New(l *zap.Logger, vars sdkservices.Vars) sdkservices.OAuth {
 			},
 
+			// Based on:
+			// https://help.salesforce.com/s/articleView?id=xcloud.remoteaccess_oauth_web_server_flow.htm&type=5
```
```suggestion
			// https://help.salesforce.com/s/articleView?id=xcloud.remoteaccess_oauth_web_server_flow.htm
```

pashafateev committed:
Update internal/backend/oauth/oauth.go

Co-authored-by: Daniel Abraham <daniel@autokitteh.com>

pashafateev committed:
address code review comments
